### PR TITLE
Clflags changes for flambda

### PR DESCRIPTION
**Pull request by mshinwell**
This pull request introduces Clflags-related changes needed for flambda.  There is also a new module Arg_helper that is used for parsing command line flag arguments for specifying optimization parameters by round (one "round" being a single run of the set of Flambda optimization passes; usually multiple rounds are used).

This has been split out separately as it is needed for the "build flambda" pull request.  There will be further changes to actually set the new Clflags options (including all the help texts etc) in a later patch.

An important thing to look at when reviewing this pull request is making sure that the existing -inline semantics has not been broken.  Note that the multiplication by the magic constant 8 has been moved to Closure.

### CODE REVIEW

**Comment by mshinwell** [comment]
One other thing: we plan to introduce a warning when Flambda flags are used in Closure mode and vice-versa.  (It was originally suggested to have an error, but @damiendoligez said that might be inconvenient when switching between modes for testing, so we now propose a warning instead.)  We will plan to implement the warning next week after the merge of the existing code has been completed.

**Comment by mshinwell** [comment]
(This also applies to more complicated specifications for -inline.  For example -inline 0=10,1=20 is legal for Flambda but always means -inline 10 for Closure.)

**Comment by mshinwell** [comment]
One other thing: there is work about to happen which will cause some of the Flambda-related compiler flags to be renamed.  I do not intend that to delay the merge of this patch.

File: utils/arg_helper.ml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+          | exception Not_found ->
+            begin match S.Value.of_string value with
+            | value -> { acc with default = value }
+            | exception exn ->
+              fatal (Printf.sprintf "%s: %s" (Printexc.to_string exn) help_text)
+            end
+          | equals ->
+            let key_value_pair = value in
+            let length = String.length key_value_pair in
+            if equals <= 0 || equals >= length - 1 then begin
```
Comment (damiendoligez) [inline utils/arg_helper.ml]: In a generic library like this one, why do you forbid the empty string? For the key, it's debatable, but for values, it's an unnecessary restriction.

File: utils/arg_helper.mli [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    (** The textual representation of a key must not contain '=' or ','. *)
+    val of_string : string -> t
+
+    module Map : Map.S with type key = t
+  end
+
+  module Value : sig
+    type t
+
+    (** The textual representation of a value must not contain '=' or ','. *)
```
Comment (damiendoligez) [inline utils/arg_helper.mli]: Why not `'='` ? AFAICT, the code allows it.

File: driver/compenv.ml [hunk 3]
```diff
@@ -205,13 +205,10 @@ let read_OCAMLPARAM ppf position =
...
-          inline_threshold := 8 * int_of_string v
-        with _ ->
-          Location.print_warning Location.none ppf
-            (Warnings.Bad_env_variable ("OCAMLPARAM",
-                                        "non-integer parameter for \"inline\""))
-        end
+      | "inline" ->
+        Float_arg_helper.parse v
+          "Bad syntax in OCAMLPARAM for 'inline'"
+          inline_threshold
```
Comment (damiendoligez) [inline driver/compenv.ml]: I'm slightly annoyed: bad syntax in `OCAMLPARAM` is normally a warning, but here you're turning it into an error.

Comment (mshinwell) [inline driver/compenv.ml]: Ack, will fix this

File: utils/arg_helper.ml (hunk 1 above)
Comment (mshinwell) [inline utils/arg_helper.ml]: We can allow empty strings if you prefer, sure.

File: utils/arg_helper.mli (hunk 2 above)
Comment (mshinwell) [inline utils/arg_helper.mli]: I will check that.

**Comment by mshinwell** [comment]
@damiendoligez Please review again, thanks.

**Comment by damiendoligez** [comment]
OK for merging.

**Pull request merged**